handlers.js

```diff
@@ -4,87 +4,113 @@ module.exports = self => {
-
-                const existingPiece = await self.find(req, props).toObject();
-
-                if (!existingPiece) {
-                  await self.insert(req, {
-                    ...props,
-                    [valueToCheck]: value
-                  });
-                } else if (!existingPiece[valueToCheck]) {
-                  const newPiece = {
-                    ...existingPiece,
-                    [valueToCheck]: value
+        return await self.apos.lock.withLock('i18n-static-lock', async () => {
+          let modified = false;
+
+          const i18nextNamespaces = Object.keys(self.apos.i18n.namespaces)
+            .filter(ns => !self.options.excludeNamespaces.includes(ns));
+
+          const plurals = {
+            plural: 'valuePlural',
+            zero: 'valueZero',
+            two: 'valuePluralTwo',
+            few: 'valuePluralFew',
+            many: 'valuePluralMany'
+          };
+
+          for (const locale of self.apos.i18n.i18next.options.languages) {
+            const req = self.apos.task.getReq({
+              locale,
+              mode: 'draft'
+            });
+            const i18nextResources = i18nextNamespaces.reduce((acc, cur) => {
+              const resources = self.apos.i18n.i18next.getResourceBundle(locale, cur);
+              return {
+                ...acc,
+                ...(resources && { [cur]: resources })
+              };
+            },
+            {});
+
+            let cachedI18nStaticResources = await self.apos.cache.get(locale, 'i18n-static');
+            if (!cachedI18nStaticResources) {
+              cachedI18nStaticResources = await self.findPiecesAndGroupByNamespace(`${locale}:draft`);
+              await self.apos.cache.set(req.locale, 'i18n-static', cachedI18nStaticResources);
+            }
+            const i18nStaticResources = cachedI18nStaticResources.reduce((acc, cur) => {
+              const ns = cur._id;
+              const resources = self.formatPieces(cur.pieces);
+              acc[ns] = resources;
+              return acc;
+            }, {});
+
+            if (!isEqual(i18nextResources, i18nStaticResources)) {
+              modified = true;
+              // eslint-disable-next-line no-console
+              console.log(`Add missing pieces in i18n-static module for ${locale}...`);
+
+              for (const [ namespace, resources ] of Object.entries(i18nextResources)) {
+                for (const [ key, value ] of Object.entries(resources || {})) {
+                  const [ title, pluralType ] = key.split('_');
+                  const valueToCheck = plurals[pluralType] || 'valueSingular';
+                  const props = {
+                    title,
+                    namespace
                   };
-                  await self.update(req, newPiece);
+
+                  const existingPiece = await self.find(req, props).toObject();
+
+                  if (!existingPiece) {
+                    await self.insert(req, {
```

Comment:
Hmm, if the situation on a new startup leads to inserts only, the `afterUpdate` handler won't fire and we won't get an updated cache and global id. But it's also probably inefficient to run that logic every time through the loop on startup anyway. So how about checking for a flag on `req` that gives a hint that this is the startup and causes `afterUpdate` to do nothing, and we factor out a more efficient implementation to run just once at the end of the startup loop if there were modifications?

Reply:
Or should I switch  `afterUpdate` to `afterSave`?

Reply:
Switched to `afterSave`